### PR TITLE
WIP: Issue #101, #89 push branch to origin

### PR DESCRIPTION
**Pull request by cchampernowne**
Created modal triggered by pushing to remote branch that doesnt exist. 

test case:
-Create remote repo
-clone the repo to your machine
-create a new local branch (not inside visualgit) 
-open repo in visualgit, make sure you are in your new branch
-Press push button
-follow prompts
-check github for existence of your new branch in your remote repo

### CODE REVIEW

**Comment by cchampernowne** [comment]
on my windows 10 machine, createUpstreamPush() isnt actually creating a remote repo. 
Visual git believes it is creating one but every time the app is restart, it seems to think the repo is only local again. 

Note: I have hardcoded values for issue 101 which could be causing interactions other users wont have

**Comment by aujlarj** [comment]
I tried testing the branch but it shows the Warning message even when there is nothing to commit or push. :-(

**Comment by katland** [comment]
I went to test this for a review and I can't log in on this branch. I enter my username and password and it just hangs on the "Signing in". I see no reason why the changes you've made would affect sign in so I'm assuming something went wrong on my end and will look into it. Just posting this on the off chance someone else goes to review this and has the same issue.
Edit: I can also no longer log in on master so it must be something on my end.

**Review by katland — CHANGES_REQUESTED** [review]
When testing I am able to get the warning error when trying to push to a repository I have only committed in and do not have a remote for. If I select the 'back' option it closes the window and continues running properly. This part is good. However if I select the 'run command' option, the application sends me back to the log in screen. (I assume the 'run command' button should be running the command without logging the user out?) From here I can repeat the same steps to get the same result:
1 - log in
2 - open the repository with only a local commit
3 - select push
4 - select run command
5 - be sent back to the log in screen to begin again
Screen Shot 2019-06-13 at 7 18 54 PM


EDIT:
These issues were resolved by updating my version of the branch with help from @smithmr 
step1 `git check master`
step2 `git pull`
step3 `git branch -D {branch name}` only do this if you have not committed anything to this branch
step4 `git checkout {branch name}`

**Does the code work? Did I test it locally? What OS am I using? Does it perform its intended function, the logic is correct etc.**
Yes, makes the test branch visible on github in my testing repository. Mac OS 10.13
**Is the code well documented? Would it be easily understood to other developers?**
yes
**Does it have correct linting/formatting?**
yes
**Is there any redundant or duplicate code? Any hard-coded values?**
no
**Can any of the code be replaced with library functions?**
no
**Are good names used for methods, variables, etc.? Do they follow naming conventions?**
yes
**Were any other issues found?**
no

**Review by katland — APPROVED** [review]

File: app/misc/git.ts [hunk 1]
```diff
@@ -1069,36 +1069,73 @@ function checkIfExistOrigin(branchName) {
...
-                return branchName;
-            });
-        });
-    });
+//push function to remote branch, we need to push to remote and then link that to the local branch
+function createUpstreamPush() {
+    //todo add loading until its successfull
+    displayModal("Pushing changes to remote... and setting upstream branch");
+  //todo update getting the branch with new function to get branch.
+  let branch = document.getElementById("branch-name").innerText;
```
Comment (mdesco18) [inline app/misc/git.ts]: Update this before merge

File: app/misc/git.ts [hunk 2]
```diff
@@ -1069,36 +1069,73 @@ function checkIfExistOrigin(branchName) {
...
+  let remoteBranch = "refs/heads/";
```
Comment (mdesco18) [inline app/misc/git.ts]: this should be refs/remotes/origin/ or the checkIfExistsOrigin() isnt working

The function still continues even on my local repository with no remotes

Suggested edit: have an error thrown if no remote repositories exist

File: app/misc/git.ts (hunk 1 above)
Comment (mdesco18) [inline app/misc/git.ts]: update this before merge

File: app/misc/git.ts [hunk 3]
```diff
@@ -1069,36 +1069,73 @@ function checkIfExistOrigin(branchName) {
...
-            return;
-        } else {
-            // tells the user if their branch is up to date or behind the remote branch
-            getAheadBehindCommits(branch).then(function (aheadBehind) {
-                if (aheadBehind.behind !== 0) {
-                    window.alert("your branch is behind remote by " + aheadBehind.behind);
-                    return;
+function pushToRemote() {
+  // checking status of remote repository and only push if you are ahead of remote
+  let branch = document.getElementById("branch-name").innerText;
```
Comment (mdesco18) [inline app/misc/git.ts]: update to new function

**Review by mdesco18 — CHANGES_REQUESTED** [review]
**Does the code work? Did I test it locally? What OS am I using? Does it perform its intended function, the logic is correct etc.** Windows 10. It works well on a repository that has a origin/remotes already existing and sets the upstreams correctly for a push when there is not one. 
**Is the code well documented? Would it be easily understood to other developers?**
very well documented
**Does it have correct linting/formatting?**
follows the formatting of similar git functions  well
**Is there any redundant or duplicate code? Any hard-coded values?** no
**Can any of the code be replaced with library functions?** no
**Are good names used for methods, variables, etc.? Do they follow naming conventions?** yes
**Were any other issues found?** aside from certain error checks, no.

File: app/misc/git.ts (hunk 1 above)
Comment (cchampernowne) [reply app/misc/git.ts]: what did you want updated here?

Comment (smithmr) [reply app/misc/git.ts]: @cchampernowne He wants the getBranch() function

**Comment by katland** [comment]
Have followed up on testing the new changes and they work on my machine.

File: app/misc/git.ts (hunk 2 above)
Comment (mdesco18) [reply app/misc/git.ts]: Suggestion Implemented

File: app/misc/git.ts (hunk 1 above)
Comment (mdesco18) [reply app/misc/git.ts]: Suggestion Implemented

File: app/misc/git.ts (hunk 3 above)
Comment (mdesco18) [reply app/misc/git.ts]: Suggestion Implemented

**Review by mdesco18 — APPROVED** [review]
**Does the code work? Did I test it locally? What OS am I using? Does it perform its intended function, the logic is correct etc.** Yes, Windows 10 tested
**Is the code well documented? Would it be easily understood to other developers?** yes
**Does it have correct linting/formatting?** yes
**Is there any redundant or duplicate code? Any hard-coded values?** no
**Can any of the code be replaced with library functions?** No, uses user functions
**Are good names used for methods, variables, etc.? Do they follow naming conventions?** Yes
**Were any other issues found?** Issues were resolved

**Review by Leon0315 — APPROVED** [review]
**Does the code work? Did I test it locally? What OS am I using? Does it perform its intended function, the logic is correct etc.** 

Tested on Mac OS 10.14 works well when I followed the test instruction. The branch created locally has successfully been pushed to remote, and the warning message is also clear.

**Is the code well documented? Would it be easily understood to other developers?**
Yes, the logic is clear and easy to follow

**Does it have correct linting/formatting?**
yes

**Is there any redundant or duplicate code? Any hard-coded values?** 
no

**Can any of the code be replaced with library functions?** 
no

**Are good names used for methods, variables, etc.? Do they follow naming conventions?** 
yes

**Were any other issues found? aside from certain error checks.**
 no.

**Pull request merged**